pytorch_lightning/plugins/checkpoint/checkpoint.py

```diff
@@ -0,0 +1,60 @@
+from abc import ABC
+from pathlib import Path
+from typing import Any, Dict, Mapping, Union
+
+from torch.nn import Module
+
+from pytorch_lightning import LightningModule
+
+
+class CheckpointPlugin(ABC):
+    def __init__(self):
+        self._training_type_plugin = None
+
+    @property
+    def training_type_plugin(self) -> "TrainingTypePlugin":
+        return self._training_type_plugin
+
+    @training_type_plugin.setter
+    def training_type_plugin(self, plugin) -> None:
+        self._training_type_plugin = plugin
+
+    @property
+    def lightning_module(self) -> LightningModule:
+        return self.training_type_plugin.lightning_module
+
+    @property
+    def model(self) -> Module:
+        return self.training_type_plugin.model
+
+    def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: str) -> None:
+        """Save model/training states as a checkpoint file through state-dump and file-write.
+
+        Args:
+            checkpoint: dict containing model and trainer state
+            filepath: write-target file's path
+        """
+
+    def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
+        """
+        Load checkpoint from a path when resuming or loading ckpt for test/validate/predict stages.
+        Args:
+            checkpoint_path: Path to checkpoint
+
+        Returns: The loaded checkpoint.
+        """
+        pass
```

Comment:
from the issue, i feel like this is the only core interface we need to support, and these ought to be abstractmethods
this way, the plugin becomes specific to IO so users can generalize beyond saving to files.

- CheckpointIO plugin -> responsible for persisting and loading the state dict
- Trainer (& training type plugin) -> responsible for gathering the overall state dict contents in-memory. the trainer uses save/load state dict hooks on the components passed to it (e.g. lightning module, data module, callbacks) in order to save and restore state as needed during the trainer execution
- Checkpoint callback -> responsible for determining 

Write path:
Checkpoint callback -> trainer.save_checkpoint -> trainer.dump_checkpoint -> trainer.training_type_plugin, callbacks, etc collect state -> trainer.training_type_plugin save_checkpoint -> calls the IO plugin's save_checkpoint 

This would avoid the training type plugin needing to be embedded in this plugin. Otherwise, this CheckpointPlugin could end up needing to know about the states for all of the other components in the overall checkpoint state dict, which will be hard to manage


One thing we need more information on is async checkpoint saving, how we determine if there are inflight checkpoints being saved, and how we do error handling in that case. @RahulMitra

Reply:
Regarding inflight checkpoints being saved, this can all be encapsulated within the checkpoint IO plugin right?

Regarding save_last, I assume this is a reference to saving twice? Any suggestions on how the plugin can assist here?